src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -44,70 +37,102 @@ public LinuxInstanceManager(IHttpClientFactory httpClientFactory, IScriptWebHost
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
+                Task.Run(async () => await HandleWarmupRequestAsync(context));
                 return true;
             }
-            else if (_assignmentContext == null)
+
+            lock (_assignmentLock)
             {
-                lock (_assignmentLock)
+                if (_assignmentContext != null)
                 {
-                    if (_assignmentContext != null)
-                    {
-                        return _assignmentContext.Equals(context);
-                    }
-                    _assignmentContext = context;
+                    return _assignmentContext.Equals(context);
                 }
+                _assignmentContext = context;
+            }
 
-                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+            _ = AssignAsync(context);
 
-                // set a flag which will cause any incoming http requests to buffer
-                // until specialization is complete
-                // the host is guaranteed not to receive any requests until AFTER assign
-                // has been initiated, so setting this flag here is sufficient to ensure
-                // that any subsequent incoming requests while the assign is in progress
-                // will be delayed until complete
-                _webHostEnvironment.DelayRequests();
+            return true;
+        }
 
-                // start the specialization process in the background
-                Task.Run(async () => await AssignAsync(context));
+        public abstract Task<string> ValidateContext(HostAssignmentContext assignmentContext);
 
-                return true;
+        private bool IsValidEnvironment(HostAssignmentContext context)
+        {
+            if (!_webHostEnvironment.InStandbyMode)
+            {
+                // This is only true when specializing pinned containers.
+                if (!context.Environment.TryGetValue(EnvironmentSettingNames.ContainerStartContext, out string startContext))
+                {
+                    _logger.LogError("Assign called while host is not in placeholder mode and start context is not present.");
+                    return false;
+                }
             }
-            else
+
+            if (_environment.IsContainerReady())
             {
-                // No lock needed here since _assignmentContext is not null when we are here
-                return _assignmentContext.Equals(context);
+                _logger.LogError("Assign called while container is marked as specialized.");
+                return false;
             }
-        }
 
-        public abstract Task<string> ValidateContext(HostAssignmentContext assignmentContext);
+            return true;
+        }
 
         private async Task AssignAsync(HostAssignmentContext assignmentContext)
         {
+            await Task.Yield();
```

Comment:
```suggestion
            await Task.Yield(); // This may be called from within a lock. This will force the lock to be released.
```

Reply:
Also a side benefit: this guarantees `StartAssignment` will be fire & forget.

Reply:
Also @fabiocav , can you confirm if that is correct? A lock will be released once a thread yields, right?

Reply:
The lock will be released, but I'm wondering if we should adjust the comment for clarity, as it isn't `Task.Yield` forcing that to happen, but the fact that when this `await` is hit, control flow will return to the caller and the lock will be released when it exits the lock scope.